cases/nbody/BMNBodyCode.hx

```diff
@@ -0,0 +1,130 @@
+@:result(-169096)
+class BMNBodyCode {
+	var bodies:Array<Body>;
+
+	public function new() {
+		this.bodies = [Body.Sun(), Body.Jupiter(), Body.Saturn(), Body.Uranus(), Body.Neptune()];
+
+		var px = 0.0;
+		var py = 0.0;
+		var pz = 0.0;
+
+		var size = this.bodies.length;
+		for (b in bodies) {
+			var m = b.mass;
+			px += b.vx * m;
+			py += b.vy * m;
+			pz += b.vz * m;
+		}
+		this.bodies[0].offsetMomentum(px, py, pz);
+	}
+
+	public function advance(dt:Float) {
+		var size = this.bodies.length;
+		for (i in 0...size) {
+			var a = bodies[i];
+			for (j in i + 1...size) {
+				var b = bodies[j];
+				var dx = a.x - b.x;
+				var dy = a.y - b.y;
+				var dz = a.z - b.z;
+
+				var distance = Math.sqrt(dx * dx + dy * dy + dz * dz);
+				var mag = dt / (distance * distance * distance);
+
+				a.vx -= dx * b.mass * mag;
+				a.vy -= dy * b.mass * mag;
+				a.vz -= dz * b.mass * mag;
+
+				b.vx += dx * a.mass * mag;
+				b.vy += dy * a.mass * mag;
+				b.vz += dz * a.mass * mag;
+			}
+		}
+
+		for (body in bodies) {
+			body.x += dt * body.vx;
+			body.y += dt * body.vy;
+			body.z += dt * body.vz;
+		}
+	}
+
+	public function energy() {
+		var e = 0.0;
+		for (i in 0...bodies.length) {
+			var a = bodies[i];
+			e += 0.5 * a.mass * (a.vx * a.vx + a.vy * a.vy + a.vz * a.vz);
+			for (j in i + 1...bodies.length) {
+				var b = bodies[j];
+				var dx = b.x - a.x;
+				var dy = b.y - a.y;
+				var dz = b.z - a.z;
+				var dist = Math.sqrt(dx * dx + dy * dy + dz * dz);
+				e -= (a.mass * b.mass) / dist;
+			}
+		}
+		return e;
+	}
+
+	public static function main() {
+		var nbodies = new BMNBodyCode();
+		for (i in 0...500000)
+			nbodies.advance(0.01);
+		Std.int(nbodies.energy() * 1000000);
```

Comment:
Is it alright to print the result here? I want to make sure that nobody optimizes away this call.

Reply:
yeah, that should be fine, output will show up in build log, but shouldn't affect measurements